syntax = "proto3";

package mavsdk.rpc.param;

import "mavsdk_options.proto";

option java_package = "io.mavsdk.param";
option java_outer_classname = "ParamProto";

// Provide raw access to get and set parameters.
service ParamService {
    /*
     * Get an int parameter.
     *
     * If the type is wrong, the result will be `WRONG_TYPE`.
     */
    rpc GetParamInt(GetParamIntRequest) returns(GetParamIntResponse) { option (mavsdk.options.async_type) = SYNC; }
    /*
     * Set an int parameter.
     *
     * If the type is wrong, the result will be `WRONG_TYPE`.
     */
    rpc SetParamInt(SetParamIntRequest) returns(SetParamIntResponse) { option (mavsdk.options.async_type) = SYNC; }
    /*
     * Get a float parameter.
     *
     * If the type is wrong, the result will be `WRONG_TYPE`.
     */
    rpc GetParamFloat(GetParamFloatRequest) returns(GetParamFloatResponse) { option (mavsdk.options.async_type) = SYNC; }
    /*
     * Set a float parameter.
     *
     * If the type is wrong, the result will be `WRONG_TYPE`.
     */
    rpc SetParamFloat(SetParamFloatRequest) returns(SetParamFloatResponse) { option (mavsdk.options.async_type) = SYNC; }
}

message GetParamIntRequest {
    string name = 1; // Name of the parameter
}

message GetParamIntResponse {
    ParamResult param_result = 1;
    int32 value = 2; // Value of the requested parameter
}

message SetParamIntRequest {
    string name = 1; // Name of the parameter to set
    int32 value = 2; // Value the parameter should be set to
}
message SetParamIntResponse {
    ParamResult param_result = 1;
}

message GetParamFloatRequest {
    string name = 1; // Name of the parameter
}
message GetParamFloatResponse {
    ParamResult param_result = 1;
    float value = 2; // Value of the requested parameter
}

message SetParamFloatRequest {
    string name = 1; // Name of the parameter to set
    float value = 2; // Value the parameter should be set to
}
message SetParamFloatResponse {
    ParamResult param_result = 1;
}

// Result type.
message ParamResult {
    // Possible results returned for param requests.
    enum Result {
        RESULT_UNKNOWN = 0; // Unknown result
        RESULT_SUCCESS = 1; // Request succeeded
        RESULT_TIMEOUT = 2; // Request timed out
        RESULT_CONNECTION_ERROR = 3; // Connection error
        RESULT_WRONG_TYPE = 4; // Wrong type
        RESULT_PARAM_NAME_TOO_LONG = 5; // Parameter name too long (> 16)
    }

    Result result = 1; // Result enum value
    string result_str = 2; // Human-readable English string describing the result
}

